atlasdb-config/src/main/java/com/palantir/atlasdb/factory/LocalOrRemoteProxy.java

```diff
@@ -63,7 +62,7 @@ protected Object handleInvocation(Object proxy, Method method, Object[] args) th
             } catch (InvocationTargetException e) {
                 Throwable targetException = e.getTargetException();
                 if (targetException instanceof NotCurrentLeaderException) {
-                    Uninterruptibles.sleepUninterruptibly(10, TimeUnit.MILLISECONDS);
+                    Thread.sleep(10);
```

Comment:
seems a bit fragile to rely on `Thread.sleep` to do the interrupted checking for you considering that there's a `while(true)` loop. If anything I'd check at the top of the while loop whether the thread is interrupted, reset the flag and then react accordingly.

Also since this is a proxy, it's weird to throw a checked exception `InterruptedException` that people can never catch (it's a compile error), especially if there is behaviour attached. Hence why your test references `UndeclaredThrowableException`. 

Tempted to say we should just throw `NotCurrentLeaderException` when we get to this case. Since that's unchecked, and you can "expect" to get that exception at some point when there is no leader for instance.

Reply:
If we're not going to propagate an InterruptedException, we should not reset the interrupt flag, otherwise the caller won't know we've been interrupted and respond appropriately.

I'm fine with throwing a different exception, but I don't think it makes much sense to propagate NotCurrentLeaderException specifically. This code is currently designed so that it never throws NotCurrentLeaderException, and it's not actually (necessarily) true that the current node isn't the leader. I think a generic illegal state exception makes the most sense (along with a set interrupted flag).

Reply:
Sure the caller won't know, however, if we don't reset the flag, if they call other blocking methods afterwards, they won't clean up, they'll just continue going. 

I suppose what I'm getting at is, making sure that if I call a method on a the proxy which doesn't throw interrupts, but needs that behaviour to be handled i.e. get out of our loop, we close properly. Then proceed to the next statement as if nothing happened.

I suppose there's also a separate piece, where we should also detect when a method is actually throwing an interrupted exception and we do the right thing there e.g. `LockService#lockWithFullResponse`.

I'm happy to take over PR to get this through considering timezones. Let me know what works for you.

Reply:
Not clear to me whether we're just talking past each other with different meanings of reset, so I'll clarify.

We need to have the interrupt flag set to true when we throw an exception (assuming we can't throw an InterruptedException), so our caller knows they were interrupted and can respond accordingly. This is what the current patch set does.

I'm fine with you taking over and finishing this.

Reply:
on second thought, the first part doesn't make sense since we need to throw, please ignore, apologies.

tested it with @jeremyk-91 and the interrupted thing is fine already, there was something along the lines of a race that could occur, where we end up throwing the wrong exception even though we *do* want to throw the `InterruptedException`., @jeremyk-91 said he'd comment so there's that.